File: Python3/1053.py
from functools import cmp_to_key
n,m,s = map(int,input().split())
weight = list(map(int,input().split()))
dic = {}
cur = []
ans = []
for _ in range(m):
    father,_,*son = map(int,input().split())
    dic[father] = son
def dfs(root,cnt,ans):
    if dic.get(root) == None and cnt - weight[root] == 0:
        cur.append(weight[root])
        new_cur = cur.copy() # list这些可变参数是直接引用的，所以如果直接apped下次cur.pop也会导致ans里的list改变
        ans.append(new_cur)
        cur.pop()
        return
    if dic.get(root) is None:
        return
    for each in dic[root]:
        cur.append(weight[root])
        dfs(each,cnt-weight[root],ans)
        cur.pop()
def mycmp(a,b):
    minlen = min(len(a),len(b))
    for i in range(minlen):
        if a[i] != b[i]:
            if a[i] > b[i]:
                return 1
            if a[i] < b[i]:
                return -1
    if len(a) > len(b):
        return 1
    if len(a) < len(b):
        return -1
    return 0

dfs(0,s,ans)
ans.sort(key=cmp_to_key(mycmp),reverse=True)
for each in ans:
    for i in range(len(each)):
        print(each[i],end='')
        if i != len(each)-1:
            print(" ",end='')
    print()

File: c++/1053.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/9/26.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <string.h>
using namespace std;

const int MAXN = 110;
int n,m,s,t,k,son;
int weight[MAXN];
vector<int> child[MAXN];
int root = 0;
vector<int> cur;
vector<vector<int>> ans;

bool cmp(vector<int> &a,vector<int> &b){
    int len = (int)min(a.size(), b.size());
    for(int i=0;i<len;++i){
        if(a[i] != b[i]) return a[i] > b[i];
    }
    return a.size() > b.size();
}

void dfs(int root,int sum){

    if(child[root].size() == 0 && sum-weight[root] == 0){
        cur.push_back(weight[root]);
        ans.push_back(cur);
        cur.pop_back();
    }

    for(auto it:child[root]){
        cur.push_back(weight[root]);
        dfs(it,sum-weight[root]);
        cur.pop_back();
    }
}

int main(){
    scanf("%d%d%d",&n,&m,&s);
    for(int i=0;i<n;++i) scanf("%d",&weight[i]);
    for(int i=0;i<m;++i){
        scanf("%d%d",&t,&k);
        for(int j=0;j<k;++j){
            scanf("%d",&son);
            child[t].push_back(son);
        }
    }
//    cur.push_back(weight[root]);
    dfs(root,s);
    sort(ans.begin(), ans.end(), cmp);
    for(auto each:ans){
        for(int i=0;i<each.size();++i){
            printf("%d",each[i]);
            if(i != each.size()-1) printf(" ");
        }
        printf("\n");
    }
}
